src/zebra_puzzles/clue_removal.py

```diff
@@ -0,0 +1,186 @@
+                }
+
+                # Check if the combination of objects and attributes are the included in the existing clue
+                if combined_obj_attributes_new.issubset(combined_obj_attributes):
+                    return True, []
+
+    elif clue_type in {"between", "not_between"}:
+        # Go through the list of chosen clues
+        for i, (clue_type_j, i_objects_j, attributes_j) in enumerate(clue_pars):
+            # Check if the new clue type and an existing clue type are a pair in redundant_clues
+            if clue_type_j == "not_same_object":
+                # Combine pairwise
+                combined_obj_attributes = {
+                    f"{x}{y}" for x, y in zip(i_objects_j, attributes_j)
+                }
+                combined_obj_attributes_new = {
+                    f"{x}{y}" for x, y in zip(clue_par[1], clue_par[2])
+                }
+
+                # Check if the combination of objects and attributes are the included in the existing clue
+                if combined_obj_attributes.issubset(combined_obj_attributes_new):
+                    if prioritise_old_clues:
+                        return True, []
+                    else:
+                        clues_to_remove.append(i)
+
+    # ---- Check if clues containing the same objects and attributes are redundant ----#
+
+    # List the clues where if the first clue is already chosen, the second clue is redundant if they contain the same objects and attributes
+    redundant_clues = {
+        ("same_object", "same_object"),
+        ("not_same_object", "not_same_object"),
+        ("left_of", "right_of"),
+        ("left_of", "not_same_object"),
+        ("right_of", "left_of"),
+        ("right_of", "not_same_object"),
+        ("just_left_of", "just_right_of"),
+        ("just_left_of", "right_of"),
+        ("just_left_of", "left_of"),
+        ("just_left_of", "next_to"),
+        ("just_left_of", "not_same_object"),
+        ("just_right_of", "just_left_of"),
+        ("just_right_of", "left_of"),
+        ("just_right_of", "right_of"),
+        ("just_right_of", "next_to"),
+        ("just_right_of", "not_same_object"),
+        ("next_to", "not_same_object"),
+        ("n_between", "not_same_object"),
+        ("between", "not_between"),
+    }
+
+    # Sort the new objects and attributes outside the following loop as they could be compared several times
+    sorted_new_objects = sorted(clue_par[1])
+    sorted_new_attributes = sorted(clue_par[2])
+
+    # Go through the list of chosen clues
+    for i, (clue_type_j, i_objects_j, attributes_j) in enumerate(clue_pars):
+        # Check if the new clue adds no new information
+        if (clue_type_j, clue_type) in redundant_clues:
+            # Check if the objects and attributes are the same
+            if (
+                sorted(i_objects_j) == sorted_new_objects
+                and sorted(attributes_j) == sorted_new_attributes
+            ):
+                return True, []
```

Comment:
Double if simplification:

```suggestion
        if (
            (clue_type_j, clue_type) in redundant_clues
            and sorted(i_objects_j) == sorted_new_objects
            and sorted(attributes_j) == sorted_new_attributes
        ):
            return True, []
```